File: src/python/redis_om_example/json_om_example.py
from redis_om import (JsonModel, EmbeddedJsonModel)
from pydantic import (PositiveInt, PositiveFloat, AnyHttpUrl)
from typing import List

# This class models the embedded "metrics" object.
class Metrics(EmbeddedJsonModel):
    rating_votes: PositiveInt
    score: PositiveFloat

# This class models the "inventory" array of objects.
class InventoryItem(EmbeddedJsonModel):
    status: str
    stock_id: str

# This class models a book.
class Book(JsonModel):
    author: str
    id: str
    description: str
    genres: List[str]
    inventory: List[InventoryItem]
    metrics: Metrics
    pages: PositiveInt
    title: str
    url: AnyHttpUrl
    year_published: PositiveInt

    # Extra configuration to specify how to generate key
    # names when saving an instance of the model in Redis.
    class Meta:
        global_key_prefix="ru204:redis-om-python"
        model_key_prefix="book"

# Create an instance of the Book model.
new_book = Book(
    author = "Redis Staff",
    id = "999",
    description = "This is a book all about Redis.",
    genres = [ "redis", "tech", "computers" ],
    inventory = [ 
        InventoryItem(
            status = "on_loan",
            stock_id = "999_1"
        ),
        InventoryItem(
            status = "maintenance",
            stock_id = "999_2"
        )
    ],
    metrics = Metrics(
        rating_votes = 4000,
        score = 4.5
    ),
    pages = 1000,
    title = "Redis for Beginners",
    url = "https://university.redis.com/courses/ru204/",
    year_published = 2022
)

# Get the locally generated ULID for this book.
print(f"new_book ULID: {new_book.pk}")

# Save the book to Redis.
new_book.save()
print("Saved book in Redis.")

# Retrieve the book from Redis.
a_book = Book.get(new_book.pk)
print("Retrieved from Redis:")
print(a_book)

# Update the author field and save it.
a_book.author = "Redis University"
a_book.save()
print("Updated author and saved to Redis:")
print(Book.get(new_book.pk))

File: src/nodejs/redis_om_example/json_om_example.js
import { Client, Entity, Schema } from 'redis-om';

const client = new Client();
await client.open(process.env.REDIS_OM_URL || 'redis://localhost:6379');

class Book extends Entity {};

const bookSchema = new Schema(Book, {
  author: { type: 'string' },
  id: { type: 'string' }, 
  description: { type: 'string' },
  genres:  { type: 'string[]' },
  pages: { type: 'number' },
  title: { type: 'string' },
  url: { type: 'string' },
  yearPublished:  { type: 'number' },
  // Redis OM Node does not yet support embedded objects, 
  // so the metrics object has been flattened to the following
  // two fields, and we have omitted the inventory array of 
  // objects for the same reason.
  ratingVotes: { type: 'number' },
  score: { type: 'number' }
}, {
  prefix: 'ru204:redis-om-node:book'
});

const bookRepository = client.fetchRepository(bookSchema);

const newBook = bookRepository.createEntity({
  author: 'Redis Staff',
  id: '999',
  description: 'This is a book all about Redis.',
  genres: [ 'redis', 'tech', 'computers' ],
  pages: 1000,
  title: 'Redis for Beginners',
  url: 'https://university.redis.com/courses/ru204/',
  yearPublished: 2022,
  ratingVotes: 4000,
  score: 4.5
});

// Get the locally generated ULID for this book.
console.log(`newBook ULID: ${newBook.entityId}`);

// Save the book to Redis.
await bookRepository.save(newBook);
console.log('Saved book in Redis.');

// Retrieve the book from Redis.
const aBook = await bookRepository.fetch(newBook.entityId);
console.log('Retrieved from Redis:');
console.log(aBook.toJSON());

// Update the author field and save it.
aBook.author = 'Redis University';
await bookRepository.save(aBook);
console.log('Updated author and saved to Redis:');
const updatedBook = await bookRepository.fetch(aBook.entityId);
console.log(updatedBook.toJSON());

await client.close();
